Update/1038 freeform block add more drawer

BoardJames opened the pull request:
This pull request adds the ability of toolbar items to have children (which allows menus) and it uses this to implement a basic more-drawer (aka kitchen sink, aka ellipse menu) for the freeform block.
![more-drawer](https://user-images.githubusercontent.com/26243021/27069198-12fd35de-5057-11e7-978e-616243b2fe17.png)

Still to do:
- ~Add arrow pointing to ellipse button as in mockup (https://github.com/WordPress/gutenberg/issues/1038 )~
- ~Fix scrolling behaviour so it does not go offscreen~

tiny-james committed:
Allow toolbar items to have child elements

tiny-james committed:
Add basic more-drawer to freeform block

tiny-james committed:
Merge branch 'master' into update/1038-freeform-block-add-more-drawer

tiny-james committed:
Fixed spelling of more-drawer

As in "cupboard drawer" not "draw a picture".

tiny-james committed:
Fixed spelling

tiny-james committed:
Added arrow to more drawer

tiny-james committed:
The more drawer expands down when there is insufficent space above

EphoxJames commented:
I have updated the more-drawer to add the arrow in the mockup and to handle the sticky behaviour better.

When there is sufficient space it displays above.
![more-drawer-above](https://user-images.githubusercontent.com/26243021/27116399-4ee38298-5114-11e7-9ee4-52163ce2e066.png)

But if you scroll down and it can't fit then it displays below.
![more-drawer-below](https://user-images.githubusercontent.com/26243021/27116419-6c62d8d2-5114-11e7-8204-9399f9b0e219.png)

tiny-james committed:
Improved styling for disabled buttons

tiny-james committed:
Merge branch 'master' into update/1038-freeform-block-add-more-drawer

youknowriad commented on blocks/library/freeform/freeform-block.js:
```diff
@@ -200,6 +250,25 @@ export default class FreeformBlock extends wp.element.Component {
+					break;
+				}
+				blockControls = blockControls.parentElement;
+			}
```
What element are we trying to get here? Could we use `element.closest()` instead of looping like this (We just need to include the polyfill `import 'element-closest';`?

EphoxJames replied:
I considered element.closest() but dismissed it because IE does not support it. I didn't know that the polyfill was available.

EphoxJames replied:
I have updated to use the polyfill.

youknowriad commented on blocks/library/freeform/freeform-block.js:
```diff
@@ -287,7 +365,19 @@ export default class FreeformBlock extends wp.element.Component {
+								<Toolbar controls={ this.mapControls( MORE_CONTROLS ) } />
+							</div>
+						),
+					} ) }
```
Should we consolidate the two `<BlockControls>` usage and the `<Fill name="Formatting.Toolbar">` into a unique `<BlockControls>`?

Something like

```js
<BlockControls>
  <Toolbar />
  <Toolbar />
  <Toolbar />
</BlockControls>
```

EphoxJames replied:
That would make things much cleaner. It should probably be done as a separate pull request though.

EphoxJames replied:
Nevermind, I see how it can be done now - I had assumed that the controls prop were the only way to set the content of the <BlockControls> element.

tiny-james committed:
Use Element.closest (with polyfill) to find the block controls

tiny-james committed:
Merge branch 'master' into update/1038-freeform-block-add-more-drawer

tiny-james committed:
Set the BlockControls content using the children prop

Previously I was using multiple calls to BlockControls using the
controls prop.

tiny-james committed:
Removed unused import

tiny-james committed:
Fixed some css issues that were not obvious on the freeform block

Added a rule to allow the toolbar css to apply to inline toolbars
Added a rule to fix colour of button text (ie subscript) when hovered

tiny-james committed:
Improved formatting

tiny-james committed:
Recalculate if the more-drawer should be below for new toolbars

tiny-james committed:
Simplified more-drawer arrow css

tiny-james committed:
Merge branch 'master' into update/1038-freeform-block-add-more-drawer
